### ISSUE TITLE
Include_role loop installs role and dependencies multiple times

### ISSUE THREAD

**Issue by Bleacks** [ansible/ansible#56046]
##### SUMMARY

I try to include roles dynamically in my project. The problem is that ansible doesn't seem to remember which role it has already installed, and hence installs every dependency, for every role, regardless of their presence in another dep or include_role.

I already saw these issues, which seemed similar at first glance but didn't help me : 
[24326](https://github.com/ansible/ansible/issues/24326)
[47374](https://github.com/ansible/ansible/issues/47374)

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

include_role

##### ANSIBLE VERSION

```paste below
ansible 2.7.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/doletma/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/doletma/.local/lib/python3.6/site-packages/ansible
  executable location = /home/doletma/.local/bin/ansible
  python version = 3.6.7 (default, Oct 22 2018, 11:32:17) [GCC 8.2.0]

```

##### CONFIGURATION

```

```
returns nothing

##### OS / ENVIRONMENT

Ubuntu 18.04.2 LTS
Testing playbook with molecule (using docker container)

##### STEPS TO REPRODUCE

Using at least 3 roles which depend on others like the following :
- role1
- role2 (dep : role1)
- role3 (dep : role1 role2)

Try to include role1 using the include_role module

```yaml
- name: Including roles dynamically
      remote_user: "{{ target_user }}"
      include_role:
        name: "{{ item }}"
      with_items: "{{ roles_to_install }}"
```

##### EXPECTED RESULTS

Installs listed roles and dependencies only once for each role

##### ACTUAL RESULTS

Installs dependencies multiple times :
- role1 : 4 times
- role2 : 2 times
- role3 : 1 time

```
TASK [Including given roles] ***************************************************
    
    TASK [role1 : debug] ***********************************************************
    ok: [localhost] => {
        "msg": "role1"
    }
    
    TASK [role1 : debug] ***********************************************************
    ok: [localhost] => {
        "msg": "role1"
    }
    
    TASK [role2 : debug] ***********************************************************
    ok: [localhost] => {
        "msg": "role2"
    }
    
    TASK [role3 : debug] ***********************************************************
    ok: [localhost] => {
        "msg": "role3"
    }
    
    TASK [role1 : debug] ***********************************************************
    ok: [localhost] => {
        "msg": "role1"
    }
    
    TASK [role2 : debug] ***********************************************************
    ok: [localhost] => {
        "msg": "role2"
    }
    
    TASK [role1 : debug] ***********************************************************
    ok: [localhost] => {
        "msg": "role1"
    }

```

I don't know if it's a bug or an intended behavior and I misunderstood / badly used, but I would like to learn how I could possibly do it 

Thank you for your time
(sorry if the issue is not properly formatted or information's missing, first time posting for help)

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/playbook/role/include.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/role/include.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jimi-c** [comment]
@Bleacks two things:

1) You should not mix include/import roles with dependencies. Using include/import allows a much finer-grained control of how roles are run versus the old way of just specifying them in the `roles:` list.
2) If you want to continue using deps, why specify all the roles in the list? Only specify the top level role and let the dependency resolution do what it's supposed to do.

That being said, I do believe this could be a bug because the role signature should not be changing in this case, as the only param passed to the role is the loop variable (`item`) with the role name in it. Could you please share your example `meta/main.yml` files for each role to ensure that the `allow_duplicates` option is not enabled for the dependent roles?

**Comment by ansibot** [comment]
@Bleacks This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jimi-c** [comment]
!needs_info

**Comment by Bleacks** [comment]
Thank you for your answer jimi.

I understand that the combined use of both include/import and role dependencies is strange, but I think there can be some context where it can be needed, for exemple :

I wanted to create a playbook, which could run every roles available inside a directory. And it seemed that one way of doing it, was to include them dynamically, based on a list of the directory content. This problem then happens if some those roles have inter dependencies.

Anyway, even if it's not a bug, I'll be glad to learn a more handsome way of doing it.

Here is the content of all meta files :
#### Role1
```
roles > role1 > meta > main.yml

---
```
#### Role 2
```
roles > role2 > meta > main.yml

---
dependencies:
  - role1
```
#### Role 3
```
roles > role3 > meta > main.yml

dependencies:
  - role1
  - role2
```

Here is the project archive if you want to run it.
[ansible-bug.zip](https://github.com/ansible/ansible/files/3406702/ansible-bug.zip)

If you need more information, please let me know

I hope it helps.
Regards

**Comment by colans** [comment]
This problem was also discussed in #22169, and summarily closed by @bcoca with:
> include_role is exempt from duplicate matching (dependencies are not, but direct use is).

@jimi-c wrote:
> You should not mix include/import roles with dependencies. Using include/import allows a much finer-grained control of how roles are run versus the old way of just specifying them in the roles: list.

I don't believe this behaviour makes sense anymore, given that users like us would like to:
1. Specify a role's dependencies, and
1. Use the newer `import_role` and `include_role` mechanisms to control how those dependencies are run.

To make this possible, I propose that we change the core behaviour so that dependencies are *only* run in advance *if* they are *not* pulled in elsewhere with `include_role` or `import_role`.

If users *do* in fact wish to run a dependency in advance with its defaults as well as with the newer mechanisms, they should run it explicitly at the beginning.

This should solve the problem once and for all of roles being run twice, once with defaults in advance (needlessly, which can also cause side effects), and then again later with proper variables set in `include_role` or `import_role`.

**Comment by gearboxscott** [comment]
Is this issue fixed yet?   we are still seeing this issue in Ansible 2.8.2, trying to see if this a coding issue on our part or it's an issue that needs fixing?

**Comment by colans** [comment]
@gearboxscott I've been tracking this issue, and haven't seen any updates anywhere about it getting fixed.  So yes, it's still open.

**Comment by peergynt** [comment]
Is this issue different from https://github.com/ansible/ansible/issues/52323?

It sounds like both issues boil down to dependent roles (i.e. meta dependency) not being de-duplicated when using `include_role` or `import_role`.

**Comment by colans** [comment]
@peergynt It's the same issue, but as they locked the thread, there's no way for us to communicate to those users that they should move over here.  Normally, I would have added a link from there to here, and mentioning it here would add a link, but not when it's locked, it seems (as you tried that already).

**Comment by kevinvalk** [comment]
Just ran into this issue as well. I dynamically include roles that have a dependency on a common role. I expected the common role to be ran once at the begining, but instead it is ran before the included role for each include.

Example
```yaml
- included_role: X
- included_role: X
```

Where the role X has a dependency set in `.meta/main.yml` like:
```yaml
dependencies:
  - role: Y
    tags: Y
```

This results in running role Y **twice**, which in my mind should only be ran **once**.

**Comment by Beefy-Swain** [comment]
Does adding the `never` tag not work to at least work around this issue?

 https://docs.ansible.com/ansible/latest/user_guide/playbooks_tags.html#special-tags

**Comment by kevinvalk** [comment]
> 
> 
> Does adding the `never` tag to the dependency declaration not work around this issue?
> 
> https://docs.ansible.com/ansible/latest/user_guide/playbooks_tags.html#special-tags

Thanks for your input, but then it is "never" ran, not even once. Except of when specifying another tag to get it included again, but then we are back to the original problem.

I think the proper solution is to avoid running the same role twice. It was what I expected from reading https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#role-duplication-and-execution

**Comment by Beefy-Swain** [comment]
Sorry, for your issue I see that that is the case, but if you just wanted to use the dependencies so that you could handle role management via galaxy, and you actually run the roles from within the parent roles, then the `never` tag has worked for me.

**Comment by droopy4096** [comment]
just tripped over this counter-intuitive behaviour. Documentation does not emphasize this aspect at all leading users into a "trap": develop "common" role on which other roles would be dependant upon (helps with heavy tag usage in playbooks and at run time as one example) only to discover that role fires *every time* despite the `allow_duplicates: false` . Even if there's no desire/capacity to fix the issue - documentation should reflect reality IMO

**Comment by snooyen** [comment]
Any movement or comment available on this?  I have a playbook that includes 3 roles [B,C,D] which all depend on the same role [A].  Role A gets executed 3 times.

`- name: myRole
  hosts: localhost
  gather_facts : false
  become: true
  tasks:
    - include_role:
        name: 'B'
        apply:
          tags: [B]
      tags: [B]
    - include_role:
        name: 'C'
        apply:
          tags: [C]
      tags: [C]
    - include_role:
        name: 'D'
        apply:
          tags: [D]
      tags: [D]
`

**Comment by niclashoyer** [comment]
We are also hitting this and I don't like to use a workaround via variables like [described here](https://github.com/ansible/ansible/issues/22169#issuecomment-429777847) because this prevents handlers from being run.

**Comment by yurtesen** [comment]
II have a similar problem

**Comment by yurtesen** [comment]
This problem seems to be going on since 2017.... :( is there any reason why it is not fixed yet?

**Comment by niclashoyer** [comment]
We worked around this issue by actually removing explicit dependencies between all roles and generating our playbooks using a template. This is not easy to handle, but we have huge playbooks that need dependencies. If every dependency would get executed multiple times the playbooks would run forever.

**Comment by yurtesen** [comment]
@niclashoyer So you implemented your own dependencies because Ansible is broken 👍 Something to consider, thanks for the suggestion.

**Comment by likan999** [comment]
Also hit by this hard and very disappointed that this fundamental issue is left untouched for years.

**Comment by Apreche** [comment]
Just somewhat recently started using ansible and was having a great time with it. Then I, as many others have, encountered this situation. Extremely disappointing and greatly reduces my confidence in the project. This is an enormous issue that seems to have been broken for many years and there is no real fix.

**Comment by aliensmily** [comment]
> Also hit by this hard and very disappointed that this fundamental issue is left untouched for years.
same here...
We are trying to write our code modular, so we got quite a few roles with dependencies and are now running into this. 
Sure there is a workaround, but it isn't a nice one.

Using dependendies in the way this issue asks for, would help to simplify code of the playbooks, when working with multiple roles.

**Comment by s-hertel** [comment]
@colans Good [comment](https://github.com/ansible/ansible/issues/56046#issuecomment-516605289).

> I don't believe this behaviour makes sense anymore, given that users like us would like to:

> 1. Specify a role's dependencies, and
> 2. Use the newer import_role and include_role mechanisms to control how those dependencies are run.

I think we've addressed this in a slightly different way than you suggested, because requiring users to run dependencies in advance explicitly would be a breaking change. In in 2.10 you can specify roles that should also be installed but not automatically run in the role's `meta/requirements.yml`. https://docs.ansible.com/ansible/latest/galaxy/user_guide.html#using-meta-requirements-yml

**Comment by s-hertel** [comment]
@Bleacks Rerunning duplicate dependencies has been fixed by https://github.com/ansible/ansible/pull/72208.

But, as pointed out by jimi-c, if you want to continue using `dependencies` in the the `meta/main.yml` file, and only run all roles a single time you should just include the parent role, role3 and let dep resolution do its thing. The behavior on devel will run role1 once as specified by the include, and once as the common dep for role2 and role3 - twice in total (as opposed to four prior to 72208). The role signature includes whether or not the role was called explicitly with an include task, so I think running it twice makes sense. But as I mentioned above, an alternative is to use the `meta/requirements.yml` so you can still manage roles with `ansible-galaxy` but avoid auto-running dependencies.

Seems like this issue is resolved?

**Comment by kousu** [comment]
I ran into this too. Here's a stripped down version of my repo that shows the problem: [ansible2.zip](https://github.com/ansible/ansible/files/6129369/ansible2.zip)

I have:


```
$ ansible-playbook --version
ansible-playbook 2.10.6
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/kousu/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.9.2 (default, Feb 20 2021, 18:40:11) [GCC 10.2.0]
```

My config has

- site.yml depends on 
  - neuropoly-base-os
    - mailer
    - neuropoly-unattended-upgrades
      - mailer
      - [jnv.unattended-upgrades](https://galaxy.ansible.com/jnv/unattended-upgrades/)
  - monitored
    - mailer
    - netdata

And indeed, I see the mailer steps happening three times; e.g. "C@example.com" gets added to ~root.forward three times.

<details><summary>full log</summary>

```
$ ansible-playbook -i hosts -l joplin --check --diff site.yml

PLAY [base] *******************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************
Enter passphrase for key '/home/kousu/.ssh/id_ed25519.neuropoly': 
ok: [joplin]

TASK [mailer : Install OpenSMTPd] *********************************************************************************************************************
The following package was automatically installed and is no longer required:
  ssl-cert
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  libasr0 libhiredis0.14 libpq5 opensmtpd-extras
The following packages will be REMOVED:
  postfix
The following NEW packages will be installed:
  libasr0 libhiredis0.14 libpq5 opensmtpd opensmtpd-extras
0 upgraded, 5 newly installed, 1 to remove and 4 not upgraded.
changed: [joplin]

TASK [mailer : Install mail(1)] ***********************************************************************************************************************
ok: [joplin]

TASK [mailer : Forward mail to sysadmins] *************************************************************************************************************
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+A@example.com

changed: [joplin] => (item=A@example.com)
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+B@example.com

changed: [joplin] => (item=B@example.com)
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+C@example.com

changed: [joplin] => (item=C@example.com)

TASK [jnv.unattended-upgrades : add distribution-specific variables] **********************************************************************************
ok: [joplin]

TASK [jnv.unattended-upgrades : add Debian Wheezy workaround] *****************************************************************************************
skipping: [joplin]

TASK [jnv.unattended-upgrades : install unattended-upgrades] ******************************************************************************************
ok: [joplin]

TASK [jnv.unattended-upgrades : install update-notifier-common] ***************************************************************************************
ok: [joplin]

TASK [jnv.unattended-upgrades : create APT auto-upgrades configuration] *******************************************************************************
--- before
+++ after: /home/kousu/.ansible/tmp/ansible-local-2127465b48_g2z8/tmpk06cogzx/auto-upgrades.j2
@@ -0,0 +1,12 @@
+// Ansible managed
+
+APT::Periodic::Unattended-Upgrade "1";
+
+APT::Periodic::Update-Package-Lists "1";
+
+
+APT::Periodic::AutocleanInterval "7";
+
+APT::Periodic::CleanInterval "7";
+
+

changed: [joplin]

TASK [jnv.unattended-upgrades : create unattended-upgrades configuration] *****************************************************************************
--- before: /etc/apt/apt.conf.d/50unattended-upgrades
+++ after: /home/kousu/.ansible/tmp/ansible-local-2127465b48_g2z8/tmplq0ri55i/unattended-upgrades.j2
@@ -1,131 +1,50 @@
-// Automatically upgrade packages from these (origin:archive) pairs
-//
-// Note that in Ubuntu security updates may pull in new dependencies
-// from non-security sources (e.g. chromium). By allowing the release
-// pocket these get automatically pulled in.
-Unattended-Upgrade::Allowed-Origins {
-	"${distro_id}:${distro_codename}";
-	"${distro_id}:${distro_codename}-security";
-	// Extended Security Maintenance; doesn't necessarily exist for
-	// every release and this system may not have it installed, but if
-	// available, the policy for updates is such that unattended-upgrades
-	// should also install from here by default.
-	"${distro_id}ESMApps:${distro_codename}-apps-security";
-	"${distro_id}ESM:${distro_codename}-infra-security";
-//	"${distro_id}:${distro_codename}-updates";
-//	"${distro_id}:${distro_codename}-proposed";
-//	"${distro_id}:${distro_codename}-backports";
+// Ansible managed
+
+// Unattended-Upgrade::Origins-Pattern controls which packages are
+// upgraded.
+Unattended-Upgrade::Origins-Pattern {
+      "origin=${distro_id},archive=${distro_codename}";
+      "origin=${distro_id},archive=${distro_codename}-security";
+      "origin=${distro_id}ESMApps,archive=${distro_codename}-apps-security";
+      "origin=${distro_id}ESM,archive=${distro_codename}-infra-security";
+      "origin=${distro_id},archive=${distro_codename}-updates";
+      "origin=${distro_id},archive=${distro_codename}-backports";
+  };
+
+// List of packages to not update (regexp are supported)
+Unattended-Upgrade::Package-Blacklist {
 };
 
-// Python regular expressions, matching packages to exclude from upgrading
-Unattended-Upgrade::Package-Blacklist {
-    // The following matches all packages starting with linux-
-//  "linux-";
-
-    // Use $ to explicitely define the end of a package name. Without
-    // the $, "libc6" would match all of them.
-//  "libc6$";
-//  "libc6-dev$";
-//  "libc6-i686$";
-
-    // Special characters need escaping
-//  "libstdc\+\+6$";
-
-    // The following matches packages like xen-system-amd64, xen-utils-4.1,
-    // xenstore-utils and libxenstore3.0
-//  "(lib)?xen(store)?";
-
-    // For more information about Python regular expressions, see
-    // https://docs.python.org/3/howto/regex.html
-};
-
-// This option controls whether the development release of Ubuntu will be
-// upgraded automatically. Valid values are "true", "false", and "auto".
-Unattended-Upgrade::DevRelease "auto";
-
-// This option allows you to control if on a unclean dpkg exit
-// unattended-upgrades will automatically run 
-//   dpkg --force-confold --configure -a
-// The default is true, to ensure updates keep getting installed
-//Unattended-Upgrade::AutoFixInterruptedDpkg "true";
 
 // Split the upgrade into the smallest possible chunks so that
-// they can be interrupted with SIGTERM. This makes the upgrade
+// they can be interrupted with SIGUSR1. This makes the upgrade
 // a bit slower but it has the benefit that shutdown while a upgrade
 // is running is possible (with a small delay)
-//Unattended-Upgrade::MinimalSteps "true";
+Unattended-Upgrade::MinimalSteps "true";
 
-// Install all updates when the machine is shutting down
-// instead of doing it in the background while the machine is running.
-// This will (obviously) make shutdown slower.
-// Unattended-upgrades increases logind's InhibitDelayMaxSec to 30s.
-// This allows more time for unattended-upgrades to shut down gracefully
-// or even install a few packages in InstallOnShutdown mode, but is still a
-// big step back from the 30 minutes allowed for InstallOnShutdown previously.
-// Users enabling InstallOnShutdown mode are advised to increase
-// InhibitDelayMaxSec even further, possibly to 30 minutes.
-//Unattended-Upgrade::InstallOnShutdown "false";
 
 // Send email to this address for problems or packages upgrades
 // If empty or unset then no email is sent, make sure that you
 // have a working mail setup on your system. A package that provides
-// 'mailx' must be installed. E.g. "user@example.com"
-//Unattended-Upgrade::Mail "";
+// 'mailx' must be installed.
+Unattended-Upgrade::Mail "root";
 
-// Set this value to one of:
-//    "always", "only-on-error" or "on-change"
-// If this is not set, then any legacy MailOnlyOnError (boolean) value
-// is used to chose between "only-on-error" and "on-change"
-//Unattended-Upgrade::MailReport "on-change";
 
-// Remove unused automatically installed kernel-related packages
-// (kernel images, kernel headers and kernel version locked tools).
-//Unattended-Upgrade::Remove-Unused-Kernel-Packages "true";
+// Do automatic removal of all unused dependencies after the upgrade
+// (equivalent to apt-get autoremove)
+Unattended-Upgrade::Remove-Unused-Dependencies "true";
 
-// Do automatic removal of newly unused dependencies after the upgrade
-//Unattended-Upgrade::Remove-New-Unused-Dependencies "true";
 
-// Do automatic removal of unused packages after the upgrade
-// (equivalent to apt-get autoremove)
-//Unattended-Upgrade::Remove-Unused-Dependencies "false";
-
-// Automatically reboot *WITHOUT CONFIRMATION* if
-//  the file /var/run/reboot-required is found after the upgrade
-//Unattended-Upgrade::Automatic-Reboot "false";
-
-// Automatically reboot even if there are users currently logged in
-// when Unattended-Upgrade::Automatic-Reboot is set to true
-//Unattended-Upgrade::Automatic-Reboot-WithUsers "true";
+// Automatically reboot *WITHOUT CONFIRMATION* if a
+// the file /var/run/reboot-required is found after the upgrade
+Unattended-Upgrade::Automatic-Reboot "true";
 
 // If automatic reboot is enabled and needed, reboot at the specific
 // time instead of immediately
 //  Default: "now"
-//Unattended-Upgrade::Automatic-Reboot-Time "02:00";
+Unattended-Upgrade::Automatic-Reboot-Time "02:00";
 
-// Use apt bandwidth limit feature, this example limits the download
-// speed to 70kb/sec
-//Acquire::http::Dl-Limit "70";
 
-// Enable logging to syslog. Default is False
-// Unattended-Upgrade::SyslogEnable "false";
 
-// Specify syslog facility. Default is daemon
-// Unattended-Upgrade::SyslogFacility "daemon";
 
-// Download and install upgrades only on AC power
-// (i.e. skip or gracefully stop updates on battery)
-// Unattended-Upgrade::OnlyOnACPower "true";
 
-// Download and install upgrades only on non-metered connection
-// (i.e. skip or gracefully stop updates on a metered connection)
-// Unattended-Upgrade::Skip-Updates-On-Metered-Connections "true";
-
-// Verbose logging
-// Unattended-Upgrade::Verbose "false";
-
-// Print debugging information both in unattended-upgrades and
-// in unattended-upgrade-shutdown
-// Unattended-Upgrade::Debug "false";
-
-// Allow package downgrade if Pin-Priority exceeds 1000
-// Unattended-Upgrade::Allow-downgrade "false";

changed: [joplin]

TASK [mailer : Install OpenSMTPd] *********************************************************************************************************************
The following package was automatically installed and is no longer required:
  ssl-cert
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  libasr0 libhiredis0.14 libpq5 opensmtpd-extras
The following packages will be REMOVED:
  postfix
The following NEW packages will be installed:
  libasr0 libhiredis0.14 libpq5 opensmtpd opensmtpd-extras
0 upgraded, 5 newly installed, 1 to remove and 4 not upgraded.
changed: [joplin]

TASK [mailer : Install mail(1)] ***********************************************************************************************************************
ok: [joplin]

TASK [mailer : Forward mail to sysadmins] *************************************************************************************************************
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+A@example.com

changed: [joplin] => (item=A@example.com)
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+B@example.com

changed: [joplin] => (item=B@example.com)
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+C@example.com

changed: [joplin] => (item=C@example.com)

TASK [neuropoly-base-os : Set hostname] ***************************************************************************************************************
--- before
+++ after
@@ -1 +1 @@
-hostname = joplin
+hostname = joplin.neuro.polymtl.ca

changed: [joplin]

TASK [neuropoly-base-os : Replace a localhost entry with our own] *************************************************************************************
--- before: /etc/hosts (content)
+++ after: /etc/hosts (content)
@@ -1,5 +1,5 @@
 127.0.0.1	localhost
-127.0.1.1	joplin
+127.0.1.1	joplin joplin.neuro.polymtl.ca
 
 # The following lines are desirable for IPv6 capable hosts
 ::1     localhost ip6-localhost ip6-loopback

changed: [joplin]

TASK [neuropoly-base-os : etckeeper] ******************************************************************************************************************
The following additional packages will be installed:
  git git-man
Suggested packages:
  git-daemon-run | git-daemon-sysvinit git-doc git-el git-email git-gui gitk
  gitweb git-cvs git-mediawiki git-svn
The following NEW packages will be installed:
  etckeeper git git-man
0 upgraded, 3 newly installed, 0 to remove and 4 not upgraded.
changed: [joplin]

PLAY [monitored] **************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************
ok: [joplin]

TASK [mailer : Install OpenSMTPd] *********************************************************************************************************************
The following package was automatically installed and is no longer required:
  ssl-cert
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  libasr0 libhiredis0.14 libpq5 opensmtpd-extras
The following packages will be REMOVED:
  postfix
The following NEW packages will be installed:
  libasr0 libhiredis0.14 libpq5 opensmtpd opensmtpd-extras
0 upgraded, 5 newly installed, 1 to remove and 4 not upgraded.
changed: [joplin]

TASK [mailer : Install mail(1)] ***********************************************************************************************************************
ok: [joplin]

TASK [mailer : Forward mail to sysadmins] *************************************************************************************************************
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+A@example.com

changed: [joplin] => (item=A@example.com)
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+B@example.com

changed: [joplin] => (item=B@example.com)
--- before: /root/.forward (content)
+++ after: /root/.forward (content)
@@ -0,0 +1 @@
+C@example.com

changed: [joplin] => (item=C@example.com)

TASK [netdata : Install netdata] **********************************************************************************************************************
The following additional packages will be installed:
  fonts-font-awesome fonts-glyphicons-halflings freeipmi-common libfreeipmi17
  libipmimonitoring6 libjs-bootstrap libjudydebian1 libnetfilter-acct1
  netdata-core netdata-plugins-bash netdata-plugins-nodejs
  netdata-plugins-python netdata-web
Suggested packages:
  freeipmi-tools apcupsd nc fping python3-psycopg2 python3-pymysql
The following NEW packages will be installed:
  fonts-font-awesome fonts-glyphicons-halflings freeipmi-common libfreeipmi17
  libipmimonitoring6 libjs-bootstrap libjudydebian1 libnetfilter-acct1 netdata
  netdata-core netdata-plugins-bash netdata-plugins-nodejs
  netdata-plugins-python netdata-web
0 upgraded, 14 newly installed, 0 to remove and 4 not upgraded.
changed: [joplin]

TASK [netdata : Configure local netdata] **************************************************************************************************************
--- before
+++ after: /home/kousu/.ansible/tmp/ansible-local-2127465b48_g2z8/tmpseejt59f/netdata.conf.j2
@@ -0,0 +1,40 @@
+# NetData Configuration
+
+# The current full configuration can be retrieved from the running
+# server at the URL
+#
+#   http://localhost:19999/netdata.conf
+#
+# for example:
+#
+#   wget -O /etc/netdata/netdata.conf http://localhost:19999/netdata.conf
+#
+
+[global]
+    run as user = netdata
+    web files owner = root
+    web files group = root
+    # Netdata is not designed to be exposed to potentially hostile
+    # networks. See https://github.com/netdata/netdata/issues/164
+    # bind socket to IP = 127.0.0.1
+    # *however*, we're behind the PolyVPN here, and it's a pain when users can't reach us, and we don't have a fulltime sysadmin team,
+    # and plus we might have streaming children trying to send metrics to us remotely,
+    # so open the ports wide by binding to all IPs, not just localhost.
+    # this could be improved by fronting with nginx?
+    bind socket to IP = 0.0.0.0
+
+    # calculator: https://learn.netdata.cloud/docs/store/change-metrics-storage#calculate-the-system-resources-ram-disk-space-needed-to-store-metrics
+    update every = 15
+    memory mode = dbengine
+    dbengine disk space = 1000
+
+
+[web]
+
+[plugin:proc:diskspace]
+    # this is necessary for some reason to convince netdata to monitor all disks,
+    # e.g. data:/srv/git/repositories would be missing otherwise
+
+    space usage for all disks = yes
+    inodes usage for all disks = yes
+

changed: [joplin]

TASK [netdata : Connect to remote netdata] ************************************************************************************************************
--- before
+++ after: /home/kousu/.ansible/tmp/ansible-local-2127465b48_g2z8/tmpobwqp68u/stream.conf.j2
@@ -0,0 +1 @@
+

changed: [joplin]

RUNNING HANDLER [netdata : Restart netdata] ***********************************************************************************************************
fatal: [joplin]: FAILED! => {"changed": false, "msg": "Could not find the requested service netdata: host"}

NO MORE HOSTS LEFT ************************************************************************************************************************************

PLAY RECAP ********************************************************************************************************************************************
joplin                     : ok=22   changed=14   unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   

```

</details>

https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#running-role-dependencies-multiple-times-in-one-playbook claims 

> Ansible only executes role dependencies once, even if defined multiple times

but that's not what I'm seeing.

I'm using the latest ansible, just uploaded yesterday, so I must have that fix, right? But it's not fixed.

**Comment by jorhett** [comment]
> @Bleacks Rerunning duplicate dependencies has been fixed by #72208.
> 
> But, as pointed out by jimi-c, if you want to continue using `dependencies` in the the `meta/main.yml` file, and only run all roles a single time you should just include the parent role, role3 and let dep resolution do its thing. With your reproducer, the behavior on devel will run role1 once as specified by the include, and once as the common dep for role2 and role3 - twice in total (as opposed to four prior to 72208). The role signature includes whether or not the role was called explicitly with an include task, so I think running it twice makes sense. But as I mentioned above, an alternative is to use the `meta/requirements.yml` so you can still manage roles with `ansible-galaxy` but avoid auto-running dependencies.
> 
> Seems like this issue is resolved?

So the main problem here is that you are referring to a usage pattern which is not documented and doesn't have examples. I think it's clear to everyone here that Ansible devs think everyone should be doing it differently, but I don't think anyone knows what that different implementation looks like.

I get that you don't want to break the old pattern. It appears you are saying that if we want to use this new pattern we have to shift all the usages in all of our roles at once (which may not be plausible for everyone)... but let's say it is. Please point to documentation that demonstrates the new usage pattern you are referring to.

I have tried moving requirements to a separate file and using `include_role` in the task list, and common dependency roles repeat ad naseum. So there's some other aspect of this style which nobody here can grasp.

**Comment by s-hertel** [comment]
@kousu That version does not include the fix. The PR I linked was merged for 2.11. Unfortunately, it was fixed via a new feature so it was not backported. The  < 2.11 documentation probably needs to be fixed because of that. The `meta/requirements.yml` is in 2.10, so it could be a workaround.

@jorhett Yeah, we need better documentation for sure. I'm not sure which part in particular you meant, but across the board more documentation, especially for roles, is a good thing.

Some [documentation](https://docs.ansible.com/ansible/latest/galaxy/user_guide.html#using-meta-main-yml) on auto-running dependencies
Some [documentation](https://docs.ansible.com/ansible/latest/galaxy/user_guide.html#using-meta-requirements-yml) on running dependencies via import/include

The [documentation](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#running-a-role-multiple-times-in-one-playbook) about rerunning roles is vague. The documentation mentions 'parameters' but doesn't list them. The examples only use 'vars' which is just one of the parameters that make up a role signature. Other parameters include `when`, `tags`, which entry points into the role are used (tasks/main.yml, vars/custom_entrypoint.yml, etc), and whether the task is called via an include (basically, anything that could change the behavior of the role and make it not behave like a "duplicate").

> I think it's clear to everyone here that Ansible devs think everyone should be doing it differently, but I don't think anyone knows what that different implementation looks like.

That isn't what I thought. It was a bug.

> I have tried moving requirements to a separate file and using include_role in the task list, and common dependency roles repeat ad naseum. So there's some other aspect of this style which nobody here can grasp.

Do you mean you moved the contents of your `meta/main.yml` to `meta/requirements.yml`? What version are you using, and if you are able to share a reproducer, that would be great.

**Comment by colans** [comment]
Okay, so it sounds like this is fixed in 2.11, but let's keep this issue open until the 2.10 documentation is updated.  That's the remaining TODO here.

This might be as simple as reverting that section to 2.9 as it sounds like it's not *actually* working in 2.10.

**Comment by jorhett** [comment]
> Do you mean you moved the contents of your `meta/main.yml` to `meta/requirements.yml`? What version are you using, and if you are able to share a reproducer, that would be great.

In my case the requirements in meta are being sourced from a monolithic repo, so there's no relevance to galaxy's requirements.  I removed them from meta, and put `include_role` in everything which previously had them in dependencies, and yet each dependent role was re-run multiple times.

This is what I mean when I say "you clearly have a vision, but nobody else knows how to use it"

**Comment by s-hertel** [comment]
@jorhett Can you open an issue with a reproducer, please

**Comment by s-hertel** [comment]
The 2.10 documentation has fixed in #75258 and tests have been added in #75259 to prevent regression. If you're having a similar issue on 2.11+ please open a new issue with reproducer. If you're struggling to figure out why a role is running multiple times, also feel free to reach out on IRC or the mailing list for help.

   * IRC: #ansible on irc.libera.chat
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
